python/sglang/srt/managers/tokenizer_manager.py

```diff
@@ -1691,6 +1733,57 @@ def _calculate_spec_decoding_metrics(
                     recv_obj.completion_tokens[i] / recv_obj.spec_verify_ct[i]
                 )
 
+    def _calculate_timing_metrics(
+        self,
+        meta_info: Dict[str, Any],
+        state: ReqState,
+        recv_obj: Union[
+            BatchStrOutput,
+            BatchEmbeddingOutput,
+            BatchMultimodalOutput,
+            BatchTokenIDOutput,
+        ],
+        i: int,
+    ) -> None:
+        """Calculate request-level timing metrics, such as inference time, decode throughput, and time per token."""
+        # Request timing timestamps.
+        if state.created_time > 0:
+            meta_info["request_received_ts"] = state.created_time
+        if state.request_scheduled_ts > 0:
+            meta_info["request_scheduled_ts"] = state.request_scheduled_ts
+        # For embeddings, there's no separate prefill phase, so omit `prefill_finished_ts`.
+        if (
+            not isinstance(recv_obj, BatchEmbeddingOutput)
+            and state.first_token_time > 0
+        ):
+            meta_info["prefill_finished_ts"] = state.first_token_time
+        if state.response_sent_ts > 0:
+            meta_info["response_sent_ts"] = state.response_sent_ts
+        if state.finished_time > 0:
+            meta_info["decode_finished_ts"] = state.finished_time
+
+        # Inference time calculation.
+        if (
+            hasattr(recv_obj, "forward_entry_time")
+            and recv_obj.forward_entry_time
+            and recv_obj.forward_entry_time[i] is not None
+            and state.finished_time_perf > 0.0
+        ):
+            forward_time = state.finished_time_perf - recv_obj.forward_entry_time[i]
+            meta_info["forward_time"] = forward_time
```

Comment:
let us call it `inference_time` or something to avoid the confusion between this and the `forward_time` in `TimeStats`?